src/reactor/segment_writer.rs

```diff
@@ -41,7 +43,10 @@ pub(crate) struct SegmentWriter {
     pub(crate) segment: ScopedSegment,
 
     /// Client connection that writes to the segmentstore.
-    connection: Option<ClientConnectionWriteHalf>,
+    pub(crate) connection: Option<ClientConnectionWriteHalf>,
+
+    /// Closes listener task before setting up new connection.
```

Comment:
yeah, perhaps it's a rare case. For example, the write method returns an error and we start to reconnect(set up new connection). But the old connection is actually still valid, in that case the listener task will not exist since the `ReadHalf` is still trying to read from segmentstore. So we have a leaked listener task.

I found this bug because I had a race condition that will set up connection twice. The `write` error first triggers a reconnect and then `read` error triggers again, leading to resend pending events twice(probably the casue of InvalidEventNumber, but I can't reproduce that error locally so I'm not sure). After fixing the race condition I thought having this logic to ensure no leaked resource is probably safer.

Reply:
Could we add a new item to the `Incoming` enum to handle it?

Reply:
Just checked the Tokio TcpStream API, the `writeHalf` can call `shutdown` to shutdown the itself, but cannot shutdown the `readHalf`. That's also why connection listener can still be running even if the `writeHalf` is dropped.